### ISSUE TITLE
Get TurnKey v16.0 building on Debian Buster

### ISSUE THREAD

**Issue by JedMeister** [turnkeylinux/tracker#1271]
Re-purposing this issue to track the current state of Core v16.0rc1 development. Obviously we're way behind schedule on this, but we're getting close to an RC1 release. This is the current state of packages and other repos for v16.0.

----
### Packages - build requirement updates
Tweaks/bugfixes to support building Buster based images on TKLDev v15.0 (i.e. build buster on stretch). These are considered legacy versions, but some will likely be updated within the lifetime of v16.x. Anything left over will be deprecated.
 - [x] autoversion (installed in buster buildroot): 
 - [x] fab https://github.com/turnkeylinux/fab/pull/5 (buster-dev has been merged into turnkey/master)
 - [x] 'turnkey' debootrap script (from TKLDev)
 - [x] turnkey-pylib (see `legacy` section)

### General
 - [x] common: <s>https://github.com/JedMeister/common/tree/buster-dev [`PR#140`](https://github.com/turnkeylinux/common/pull/140)</s> Merged into master.
    - major changes for buster
 - [x] cdroots: <s>https://github.com/JedMeister/cdroots/tree/buster-dev [`PR#2`](https://github.com/turnkeylinux/cdroots/pull/2)</s> Merged into master.
 - [x] bootstrap: https://github.com/turnkeylinux/bootstrap/pull/2 Merged into master.
 - [x] buildroot: https://github.com/turnkeylinux/buildroot/pull/2 Merged into master.
 - [x] core: https://github.com/JedMeister/core/tree/buster-dev [`PR#7`](https://github.com/turnkeylinux-apps/core/pull/7)
    - only changelog so far...

### Packages - updates
Major updates for v16.0 - many are transitions to python3. (Also not sure why this section has funky formatting?!)
 - [x] di-live: <s>https://github.com/JedMeister/di-live/tree/buster-dev [`PR#7`](https://github.com/turnkeylinux/di-live/pull/7)</s> Merged into master.
    - py3; major refactoring
 - [x] confconsole: <s>https://github.com/JedMeister/confconsole/tree/python3 [`PR#28`](https://github.com/turnkeylinux/confconsole/pull/28)</s> Merged into master.
    - py3 update and refactoring 
 - [x] inithooks: <s>https://github.com/JedMeister/inithooks/tree/python3 [`PR#34`](https://github.com/turnkeylinux/inithooks/pull/34)</s> Merged into master.
    - py3 update and refactoring
 - [x] turnkey-sysinfo: <s>https://github.com/JedMeister/turnkey-sysinfo/tree/python3 [`PR#3`](https://github.com/turnkeylinux/turnkey-sysinfo/pull/3)</s> Merged into master.
    - py3 update; now also includes turnkey-version script
 - [x] turnkey-version: <s>https://github.com/JedMeister/turnkey-version/tree/transitional-package [`PR#3`](https://github.com/turnkeylinux/turnkey-version/pull/3)</s> Merged into master.
    - transitional package; previous functionality merged into turnkey-sysinfo

 - [x] webmin-tklbam: <s>https://github.com/JedMeister/webmin-tklbam/tree/buster-dev [`PR#1`](https://github.com/turnkeylinux/webmin-tklbam/pull/1)</s> Merged into master.
    - minor update to remove build warnings
 - [x] webmin: <s>[`PR#3`](https://github.com/turnkeylinux/webmin/pull/3)</s> Merged into master; tagged as appropriate.
    - update to upstream version v1.930 plus improved packaging and inclusion of systemd service file

### Packages - new
New for v16.0 - most pulled from `turnkey-pylib`
 - [x] turnkey-stdtrap: https://github.com/turnkeylinux/turnkey-stdtrap
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-command: https://github.com/turnkeylinux/turnkey-command
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-paths: https://github.com/turnkeylinux/turnkey-paths
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-lazyclass: https://github.com/turnkeylinux/turnkey-lazyclass
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-popen4: https://github.com/turnkeylinux/turnkey-popen4
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-conffile: https://github.com/JedMeister/turnkey-conffile [`PR#1`](https://github.com/turnkeylinux/turnkey-conffile/pull/1)
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-temp: https://github.com/turnkeylinux/turnkey-temp
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-fifobuffer: https://github.com/turnkeylinux/turnkey-fifobuffer
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-fileevent: https://github.com/turnkeylinux/turnkey-fileevent
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-paged: https://github.com/turnkeylinux/turnkey-paged
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-pidlock: https://github.com/turnkeylinux/turnkey-pidlock
    - pulled from turnkey-pylib & updated to python3
 - [x] turnkey-netinfo: https://github.com/turnkeylinux/turnkey-netinfo
    - pulled from turnkey-pylib & updated to python3

### Packages - legacy
tklbam and dependencies; many unchanged (just rebuilt against buster). Still need to be updated to python3 ASAP but postponed so as not to block v16.0 release. With the exception of tklbam, all these will be deprecated once they are no longer required (will depend on default Debian packages and/or have new requirements packaged separately).
 - [x] tklbam: https://github.com/turnkeylinux/tklbam
    - unchanged
 - [x] tklbam-duplicity: https://github.com/turnkeylinux/tklbam-duplicity
    - unchanged
 - [x] tklbam-python-boto: https://github.com/turnkeylinux/tklbam-python-boto
    - unchanged
 - [x] tklbam-squid: https://github.com/turnkeylinux/tklbam-squid
    - unchanged
 - [x] turnkey-pylib: <s>https://github.com/JedMeister/turnkey-pylib/tree/buster-dev [`PR#2`](https://github.com/turnkeylinux/turnkey-pylib/pull/2)</s>
    - modifications to support 'build requirement updates' and/or tklbam

----

Some other links of interest:
https://wiki.debian.org/DebianBuster
https://wiki.debian.org/NewInBuster
https://www.debian.org/releases/buster/

**Comment by JedMeister** [comment]
Furthermore, during this update, we should probably be "future-proofing" as much as possible and adding in build support for future Debian releases (e.g. it's not really that big a deal to add in potential support for "Debian 11/Bullseye" and perhaps even "Debian 11/Bookworm"?!)

**Comment by kaktux** [comment]
sign me up for trying this out.
I actually just updated turnkey 15 on a vm i want to use for webpage development (wordpress -> syncing from vm to staging+live via wordmove (needing ruby 2.4+)) myself by simply editing the sources.list as i needed a higher ruby version. Beeing new to ruby trying to install via rvm or rbenv turned out to be a pain in the... and coming with errors. So as debian 9 offers ruby 2.3 via simply adding the package ruby-full and debian 10 uses 2.5 - it was much faster simply to update the whole thing then to solve the errors using rvm and rbenv.

Update - worked without problems. 
The only thing broken now is that i can no longer access the turnkey menu in the vm (it goes straight to console - which for some reason has qwerty keyboard layout (which btw is an "error" i have while installing every turnkey version - and can only be solved after finishing the install by simply changing keyboard layout), package shellinabox (which doesnt ring a bell for me before) crashed during the update (which is easily solved by not using the browser shell (port 12320) = shellinabox... but the terminal via virtualbox).
The whole thing runs without problems/crashes

Long story short: it is fine for a deveolpment vm like now - i dont really need the broken stuff: but - if you have a debian 10 iso/img to work with - i would gladly give it a try.
And i also agree - as debian 9 will come to an end in the near future it probably is a "waste" to work on updating those turnkey instances - and focus on the debian 10 based instead - especially with the timetable you mentioned.
Maybe even play around with 2 core versions at a time - based on stable + unstable.

**Comment by JedMeister** [comment]
Hey @kaktux - Apologies on my slow response. I actually thought I'd already responded, but I must not have hit save (or "comment" to be precise).

It turned out that we possibly jumped the gun a little as we had masses of issues with the Buster installer and it was hard to work out which were our issues and which were Debian issues. Also, the approach that we took to get it going was a bit ad hoc and wasn't as methodical as it might have been (I take full responsibility for that)...

Thanks tons for your testing and feedback. It's really valuable to get those insights.

FWIW, we started at the other end (building our custom software for Buster and working on the install process) and after pouring quite a few resources into it and hitting so many issues, we decided to back out for the moment (things are almost certainly better with Debian by now).

We're currently regrouping and forming a more coherent attack plan. The unfortunate reality of that is that it means there will be further delays (e.g. we're going to update all our Python code to Python3 before we even begin relooking at the Buster build).

I will certainly keep you in mind though and would love to have some assistance testing. There won't be any "official" ISOs until we get to an RC of Core. The way it went with Stretch/v15.0 development was that I published documentation on how to build ISOs using [TKLDev](https://www.turnkeylinux.org/tkldev) and those assisting with the testing built their own. However, if there is interest, I could probably upload "development" ISOs to a personal AWS bucket somewhere for yourself and anyone else who's interested.

Please feel free to bump if you would like an update. I still hope to have something close by the time Buster is released but I'm not 100% sure how realistic that is. Although I definitely plan to do better than we did with Stretch - IIRC it was almost 6mths after Stretch became stable that we released the first 15.0RC!

**Comment by JedMeister** [comment]
FWIW, it looks like we're still a little while away from Buster stable. At the time of writing there are still 115 unaddressed RC (release critical in this context) bugs. You can view the current bug stats for testing [here](https://bugs.debian.org/release-critical/other/testing.html).

**Comment by JedMeister** [comment]
Closing this now as we've [released rc versions of Core and TKLDev](https://www.turnkeylinux.org/blog/v16.0rc1-core-and-tkldev) and hopefully will start releasing stable appliances really soon.
